httpclient5/src/main/java/org/apache/hc/client5/http/entity/compress/CommonsCompressDecoderFactory.java

```diff
@@ -0,0 +1,122 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+
+package org.apache.hc.client5.http.entity.compress;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Locale;
+
+import org.apache.commons.compress.compressors.CompressorException;
+import org.apache.commons.compress.compressors.CompressorStreamFactory;
+import org.apache.hc.client5.http.entity.InputStreamFactory;
+import org.apache.hc.core5.annotation.Contract;
+import org.apache.hc.core5.annotation.Internal;
+import org.apache.hc.core5.annotation.ThreadingBehavior;
+
+/**
+ * {@link InputStreamFactory} backed by
+ * <a href="https://commons.apache.org/proper/commons-compress/">Apache&nbsp;Commons Compress</a>.
+ * <p>
+ * The class is compiled against Commons Compress but lives behind an <i>optional</i>
+ * dependency.  At run-time it will only be loaded when the library is present,
+ * therefore callers may rely on it without pulling Commons Compress into every
+ * downstream build.
+ * </p>
+ *
+ * <h4>Run-time guards</h4>
+ * Some encodings (e.g.&nbsp;{@code br}, {@code zstd}, {@code xz/lzma})
+ * depend on native helper JARs.  {@link #runtimeAvailable(String)} performs a
+ * lightweight {@code Class.forName} probe so we register such codecs only when
+ * the helper is on the class-path.
+ *
+ * @since 5.6
+ */
+@Internal
+@Contract(threading = ThreadingBehavior.STATELESS)
+final class CommonsCompressDecoderFactory implements InputStreamFactory {
+
+    private final String encoding;
+
+    CommonsCompressDecoderFactory(final String encoding) {
+        this.encoding = encoding.toLowerCase(Locale.ROOT);
+    }
+
+    @Override
+    public String getContentEncoding() {
+        return encoding;
+    }
+
+    @Override
+    public InputStream create(final InputStream source) throws IOException {
+        try {
+            return new CompressorStreamFactory()
+                    .createCompressorInputStream(encoding, source);
+        } catch (final CompressorException | LinkageError ex) {
+            throw new IOException(
+                    "Unable to decode Content-Encoding '" + encoding + '\'', ex);
+        }
+    }
+
+    private enum Probe {
```

Comment:
This feels a bit redundant with the new `ContentCoding` enum. What do you think about moving the "requires this type name" aspect (a.k.a. "probe class") of `Probe` to `ContentCoding` and turning this "Probe" into a `Set<ContentCoding>`?